openmdao/components/input_resids_comp.py

```diff
@@ -0,0 +1,126 @@
+        ----------
+        name : str
+            Name of the variable in this component's namespace.
+        val : float or list or tuple or ndarray or Iterable
+            The initial value of the variable being added in user-defined units.
+            Default is 1.0.
+        shape : int or tuple or list or None
+            Shape of this variable, only required if val is not an array. Default is None.
+        units : str or None
+            Units in which this input variable will be provided to the component
+            during execution. Default is None, which means it is unitless.
+        desc : str
+            Description of the variable.
+        tags : str or list of strs
+            User defined tags that can be used to filter what gets listed when calling
+            list_inputs and list_outputs.
+        shape_by_conn : bool
+            If True, shape this input to match its connected output.
+        copy_shape : str or None
+            If a str, that str is the name of a variable. Shape this input to match that of
+            the named variable.
+        compute_shape : function
+            A function taking a dict arg containing names and shapes of this component's outputs
+            and returning the shape of this input.
+        distributed : bool
+            If True, this variable is a distributed variable, so it can have different sizes/values
+            across MPI processes.
+        ref : float or ndarray or None
+            Scaling parameter. The value in the user-defined units of this residual
+            when the scaled value is 1. Default is 1.
+        """
+        self._refs[name] = ref
+        super().add_input(name, val=val, shape=shape, units=units, desc=desc, tags=tags,
+                          shape_by_conn=shape_by_conn, copy_shape=copy_shape,
+                          compute_shape=compute_shape, distributed=distributed)
+
+    def setup_residuals(self):
+        """
+        Delay calls for add_residual for this component.
+
+        This method is used since input/residual sizes may not
+        be known until final setup.
+        """
+        for name in self._var_rel_names['input']:
+            meta = self._var_rel2meta[name]
+            resid_name = f'resid_{name}'
+            self.add_residual(resid_name, shape=meta['shape'], units=meta['units'],
+                              desc=meta['desc'], ref=self._refs[name])
+
+    def setup_partials(self):
+        """
+        Delay calls to declare_partials for the component.
+
+        This method is used because input/residual sizes
+        may not be known until final setup.
+        """
+        size = self._get_var_meta('x', 'size')
+        self.mat = np.eye(size) * 3.
+        rng = np.arange(size)
+        self.declare_partials('y', 'x', rows=rng, cols=rng, val=3.0)
+        for name in self._var_rel_names['input']:
+            resid_name = 'resid_' + name
+            size = self._get_var_meta(name, 'size')
```

Comment:
Since you already have the relative name, it would be quicker to just get size from self._var_rel2meta[name]['size']